fix(runtimes): propagate Trainer.NumNodes into TemplateSpec (Parallelism/Completions)

NarayanaSabari opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR fixes the `RuntimeInfo` helpers to correctly synchronize `PodSets.Count` values (set by `numNodes`) to the `TemplateSpec.ObjApply` (JobSet specification).

Previously, when external consumers like Kueue called `RuntimeInfo()`, the `info.TemplateSpec.PodSets[*].Count` was correctly updated by ML policy plugins (Torch, PlainML, MPI), but the underlying JobSet spec (`info.TemplateSpec.ObjApply`) still had the original `Parallelism` and `Completions` values. This required consumers to implement post-processing workarounds.

**Changes:**
1. Added `SyncPodSetsToTemplateSpec()` method in `runtime.go` that syncs `PodSets.Count` to JobSet's `Parallelism`/`Completions`
2. Called this method in `RuntimeInfo()` after `RunEnforceMLPolicyPlugins()` runs
3. Added unit tests for the new sync function

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubeflow/trainer/issues/3042

Related Kueue issue: https://github.com/kubernetes-sigs/kueue/issues/8287

#### Special notes for your reviewer:

- The existing comment in `runtime.go` line 45 already mentioned: "ObjApply podSpecs and this PodSets should be kept in sync by info.SyncPodSetsToTemplateSpec()" - this PR implements that function.
- Once this is merged, the workaround in Kueue's `trainjob_controller.go` can be removed.
- All existing tests pass.

NarayanaSabari committed:
Fix RuntimeInfo helpers to correctly sync numNodes to TemplateSpec

- Add SyncPodSetsToTemplateSpec() method to sync PodSets.Count to JobSet Parallelism/Completions
- Call sync after EnforceMLPolicyPlugins run in RuntimeInfo()
- Add unit tests for SyncPodSetsToTemplateSpec

Fixes #3042

Signed-off-by: narayanasabari <sabarinarayanakg@proton.me>

kaisoz commented on pkg/runtime/runtime.go:
```diff
@@ -196,6 +196,32 @@ func TemplateSpecApply[A any](info *Info) (*A, bool) {
+// the underlying template spec (e.g., JobSet's Parallelism/Completions) reflects those changes.
+// This is necessary for external consumers of RuntimeInfo (like Kueue) who may not call Build().
+// Matching is done by name for robustness against ordering differences.
+func (i *Info) SyncPodSetsToTemplateSpec() {
```
IMO, the `info` object shouldn't be handling this. It is meant to be an abstract structure used during the orchestration of the TrainJob, allowing for other runtimes in the future that might not be JobSet-based (e.g., SlurmRuntime). This means it should stay as agnostic as possible regarding JobSets (please maintainers correct me if I'm wrong here).

I think this logic should be **kept at the plugin level but moved to the JobSet object builder**. Currently, the assignment is [performed in the jobset plugin](https://www.google.com/search?q=link) before calling the JobSet Builder.

My proposal is to add a function to the [jobset builder](https://www.google.com/search?q=link) (which is what Kueue calls). Maybe something like PodSets(ps []runtime.PodSet) *Builder? What do you think?

cc @astefanutti @andreyvelich @tenzen-y

kaisoz replied:
Hey Sabari! Sorry for this late reply, I've been away for some time

> 1. The existing comment at line 45 already mentioned info.SyncPodSetsToTemplateSpec() as the intended sync mechanism, suggesting this was part of the original design

I missed that! To be fair, the `Builder`change was a transitional solution given that there's an ongoing discussion about what is missing in the `Info` object so that external controllers (like Kueue) don't need to recreate the JobSet object. Besides, Kueue already calls the Builder, which means it has knowledge of the JobSet creation mechanisms.

> 2. The TemplateSpecApply generic function already does runtime-specific type casting from Info, so this follows a similar pattern

Not exactly...  `TemplateSpecApply` is a generic function where the caller specifies the type so that `runtime.go` itself doesn't import `JobSet` types. But `SyncPodSetsToTemplateSpec` uses `JobSetSpecApplyConfiguration` directly, creating a coupling we're trying to avoid

With this in mind, I gave it a second thought and have a new proposal. Since the core issue is that 
`SyncPodSetsToTemplateSpec` on `Info` couples it to `JobSet`, we can leverage the existing framework plugin pattern to keep that separation clean:

1. Create a a new interface in `runtime/framework/interface.go` E.g:

```
type TemplateSpecSyncPlugin interface {
    Plugin
    SyncPodSetsToTemplateSpec(info *runtime.Info) error
}
```

2. Implement such function in the `jobset` plugin (`plugins/jobset/jobset.go`) 
3. Add a new function to `framework/core/framework.go` which calls the plugin sync function
4. Call that function from the `TrainingRuntime` as with the rest (e.g `RunEnforceMLPolicyPlugins`)

In this way, you keep `runtime.Info` truly template-agnostic while placing the sync responsibility in the plugin that understands the underlying template type. WDYT?

andreyvelich replied:
I agree with @kaisoz, we should not introduce job specific code into runtime framework.
We need to find a way to sync some data from plugins back to the underlying job, like numNodes, resources Per node, etc. For example, we should delegate assignment of container resources to MPI plugin instead of doing it in the JobSet builder: https://github.com/kubeflow/trainer/blob/26bf9cdf7b8c6ba0e8e0d248462ed78bc1c29679/pkg/runtime/framework/plugins/jobset/builder.go#L143-L144

With your approach @kaisoz, how this template syncer can be invoked from the plugins (e.g. Torch, PlainML, MPI) ?

NarayanaSabari replied:
Thanks @andreyvelich and @kaisoz for the detailed architectural feedback!

I understand the concern about coupling `Info` to JobSet-specific types. The plugin framework approach makes sense to maintain runtime-agnostic design.

After analyzing the codebase deeply, I have answers to the architecture questions, but one critical clarification is needed:

## My Analysis:

### 1. Current Plugin Flow:
```go
// In TrainingRuntime.RuntimeInfo():
info, err := r.newRuntimeInfo(trainJob, jobSetTemplateSpec, mlPolicy, podGroupPolicy)
↓
r.framework.RunEnforceMLPolicyPlugins(info, trainJob)  
// ML plugins (Torch/MPI/PlainML) modify info.TemplateSpec.PodSets[*].Count
↓
info.SyncPodSetsToTemplateSpec() // CURRENT: Couples Info to JobSet
↓
r.framework.RunEnforcePodGroupPolicyPlugins(info, trainJob)
↓
r.framework.RunPodNetworkPlugins(info, trainJob)
```

### 2. Sync Happens AFTER ML Policy Plugins
The sync must occur **after** `RunEnforceMLPolicyPlugins` because ML plugins modify `PodSets.Count` values:
- **Torch**: Line 105 sets `*trainerPS.Count = *trainJob.Spec.Trainer.NumNodes`
- **MPI**: Line 117 sets `*node.Count = max(*trainJob.Spec.Trainer.NumNodes-1, 1)`
- **PlainML**: Line 54 sets `*trainerPS.Count = *trainJob.Spec.Trainer.NumNodes`

### 3. JobSet Plugin Already Syncs in Build()
The **JobSet plugin's `Build()` method** (lines 262-264 in `jobset.go`) **already syncs** `PodSets.Count` to `Parallelism`/`Completions`:
```go
for psIdx, ps := range info.TemplateSpec.PodSets {
    if ps.Count != nil {
        jobSetSpec.ReplicatedJobs[psIdx].Template.Spec.Parallelism = ps.Count
        jobSetSpec.ReplicatedJobs[psIdx].Template.Spec.Completions = ps.Count
    }
```

**The problem:** External consumers like Kueue call `RuntimeInfo()` directly (not `Build()`), so they never get the synced values.

## Proposed Solution:

Following @kaisoz's plugin framework pattern:

```go
// 1. runtime/framework/interface.go - New plugin interface
type TemplateSpecSyncPlugin interface {
    Plugin
    SyncPodSetsToTemplateSpec(info *runtime.Info) error
}

// 2. plugins/jobset/jobset.go - Implement in JobSet plugin
func (j *JobSet) SyncPodSetsToTemplateSpec(info *runtime.Info) error {
    jobSetSpec, ok := runtime.TemplateSpecApply[jobsetv1alpha2ac.JobSetSpecApplyConfiguration](info)
    if !ok || jobSetSpec == nil {
        return nil
    }
    
    for _, ps := range info.TemplateSpec.PodSets {
        if ps.Count == nil {
            continue
        }
        // Find the matching ReplicatedJob by name (name-based matching for robustness)
        for rJobIdx := range jobSetSpec.ReplicatedJobs {
            if jobSetSpec.ReplicatedJobs[rJobIdx].Name != nil && 
               *jobSetSpec.ReplicatedJobs[rJobIdx].Name == ps.Name {
                jobSetSpec.ReplicatedJobs[rJobIdx].Template.Spec.Parallelism = ps.Count
                jobSetSpec.ReplicatedJobs[rJobIdx].Template.Spec.Completions = ps.Count
                break
            }
        }
    }
    return nil
}

// 3. framework/core/framework.go - Add framework orchestration
func (f *Framework) RunTemplateSpecSyncPlugins(info *runtime.Info) error {
    for _, plugin := range f.templateSpecSyncPlugins {
        if err := plugin.SyncPodSetsToTemplateSpec(info); err != nil {
            return err
        }
    }
    return nil
}

// 4. runtime/core/trainingruntime.go - Call after ML policy plugins
func (r *TrainingRuntime) RuntimeInfo(...) (*runtime.Info, error) {
    // ... existing code ...
    if err = r.framework.RunEnforceMLPolicyPlugins(info, trainJob); err != nil {
        return nil, err
    }
    
    // NEW: Sync after ML plugins modify counts
    if err = r.framework.RunTemplateSpecSyncPlugins(info); err != nil {
        return nil, err
    }
    
    if err = r.framework.RunEnforcePodGroupPolicyPlugins(info, trainJob); err != nil {
        return nil, err
    }
    // ... rest of code ...
}
```

## Question 

**Should the JobSet plugin's `Build()` method also use the new `SyncPodSetsToTemplateSpec()` to avoid code duplication?**

Currently `Build()` has its own sync logic (lines 262-264). We could:
- **Option A:** Keep both (Build has its own, plugin interface has separate implementation)
- **Option B:** Make Build() call `j.SyncPodSetsToTemplateSpec(info)` to reuse logic

Once this is clarified, I'll implement the full refactor with tests.

mimowo replied:
Folks, this PR looks stale on the question. 

Can we have some guidance for the contribution so that it can move forward? It would be great to offload Kueue code from this hack at our side (and something all other frameworks would need to do, so currently this is very error prone).

cc @tenzen-y @andreyvelich @astefanutti

andreyvelich commented:
/ok-to-test

NarayanaSabari committed:
refactor: move SyncPodSetsToTemplateSpec into BuildParallelCountPlugin

Move the PodSets-to-TemplateSpec synchronization logic out of
runtime.Info and into the plugin framework as BuildParallelCountPlugin.
This keeps runtime.Info runtime-agnostic and places JobSet-specific
logic in the jobset plugin where it belongs.

- Add BuildParallelCountPlugin interface to framework/interface.go
- Implement BuildParallelCount on JobSet plugin with name-based matching
- Register plugin in Framework.New() and add RunBuildParallelCountPlugins
- Update trainingruntime.go to call the plugin instead of info method
- Move tests from runtime_test.go to jobset_test.go with additional cases

Signed-off-by: Sabari Narayana <sabarinarayanakg@proton.me>

kaisoz commented:
> > Instead of placing SyncPodSetsToTemplateSpec() on runtime.Info, I've introduced a new BuildParallelCountPlugin interface in the plugin framework. This keeps runtime.Info runtime-agnostic and moves the JobSet-specific syncing logic into the jobset plugin where it belongs.
> 
> Do we want to build a separate plugin for every change that needs to be added to the JobSet spec from other plugins?

@NarayanaSabari From this [comment](https://github.com/kubeflow/trainer/pull/3057#discussion_r2782588415)

> ....We probably can start to implement BuildParallelCount in ComponentBuilderPlugin

I think what @tenzen-y  meant was to have an additional function `BuildParallelCount` in the [ComponentBuilderPlugin interface](https://github.com/kubeflow/trainer/blob/abf42d5dab50b306e0f9ec0c69790c8cc5ef81f7/pkg/runtime/framework/interface.go#L61) which would be called before the `Build` command  [in the framework](https://github.com/kubeflow/trainer/blob/abf42d5dab50b306e0f9ec0c69790c8cc5ef81f7/pkg/runtime/framework/core/framework.go#L138)

And actually, I do think it's a good idea to have an additional function in the `ComponentBuilderPlugin` instead of a new plugin interface. However, I still have preference for a single function instead of a set of specialized functions.. WDYT?

cc @andreyvelich
